### ISSUE TITLE
Merge concat fails if only some of the inputs are masked

### ISSUE THREAD

**Issue by pdasigi** [fchollet/keras#3205]
Please make sure that the boxes below are checked before you submit your issue. Thank you!

- [X] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [X] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [X] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

Here is the code that reproduces the issue. Only the last line throws an error.

```
from keras.layers import Input, Embedding, merge

input1 = Input(shape=(2,), dtype='int32')
input2 = Input(shape=(2,), dtype='int32')
embedding = Embedding(input_dim=5, output_dim=10)
masked_embedding = Embedding(input_dim=5, output_dim=10, mask_zero=True)
embedded_input1 = embedding(input1)
embedded_input2 = embedding(input2)
masked_embedded_input1 = masked_embedding(input1)
masked_embedded_input2 = masked_embedding(input2)
both_unmasked_concat = merge([embedded_input1, embedded_input2], mode='concat')
both_masked_concat = merge([masked_embedded_input1, masked_embedded_input2], mode='concat')
one_masked_sum = merge([masked_embedded_input1, embedded_input2], mode='sum')
one_masked_concat = merge([masked_embedded_input1, embedded_input2], mode='concat')
```

I believe the issue is with this line: https://github.com/fchollet/keras/blob/master/keras/engine/topology.py#L1334
I guess @codekansas meant it to be
`masks = [K.ones_like(i[:-1]) if m is None else m for i, m in zip(inputs, mask)]`
but I'm not sure why there's `[:-1]`

**Comment by codekansas** [comment]
See the issue here for a fix https://github.com/fchollet/keras/issues/3065

**Comment by pdasigi** [comment]
Thanks for the pointer!

**Comment by pdasigi** [comment]
This PR fixes the issue: https://github.com/fchollet/keras/pull/3218. Closing it.

**Comment by nelson-liu** [comment]
@codekansas @braingineer @pdasigi looks like this bug wasn't completely fixed by #3218 --- the example in the original post still breaks on the TF backend with the error:
```
TypeError: Tensors in list passed to 'values' of 'ConcatV2' Op have types [bool, uint8] that don't all match.
```

I think the reason this happens is because #3218 explicitly casts unmasked things to `uint8` (see [this line of code](https://github.com/fchollet/keras/blob/master/keras/engine/topology.py#L1525-L1527)), but the embedding layer provides a `bool` mask in TF (see [`compute_mask` in embedding.py](https://github.com/fchollet/keras/blob/master/keras/layers/embeddings.py#L106-L109)). 

That being said, it seems like a cast would fix the issue but i'm unclear on whether we should be casting `uint8`to `bool` in `topology.py` or `bool` to `uint8` in `compute_mask` of `embeddings.py`. Any pointers?
